walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,101 @@
+    meeting. Sort all of events in descending order by end time. Grab the start
+    time for this current event and traverse through array all events that have
+    an end time greater than this start time. The time slot that can not be used
+    is then the end time of the current iteration and the minimum start time of
+    those other events. Now take the event that after the last event that was
+    discovered still within the constraints and repeat from there. Such an
+    approach will take O(nlogn) time to sort and O(n) time to iterate. The
+    removal will take O(mn) time giving us a complexity of O(mn
+    + nlogn).
+    */
+
+    // Store the times when we cannot meet, essentially take the inverse of this
+    // to determine the times we can meet.
+
+    Collection<String> meetingAttendees = request.getAttendees();
+    ArrayList<TimeRange> fullTimes = new ArrayList<>();
+    Collection<TimeRange> toReturn = new ArrayList<>();
+    // Change to ArrayList so we can sort.
+    ArrayList<Event> eventList = new ArrayList<>(events);
+    // Remove an event if none of the members in the meeting attended it.
+    eventList.removeIf(event
+                       -> !event.getAttendees().stream().anyMatch(
+                           attendee -> meetingAttendees.contains(attendee)));
+    Collections.sort(eventList, ORDER_END.reversed());
+
+    ListIterator<Event> it = eventList.listIterator();
+    while (it.hasNext()) {
+      Event currentEvent = it.next();
+      TimeRange currentTime = currentEvent.getWhen();
+      int currentEndTime = currentTime.end();
+      int currentStartTime = currentTime.start();
+
+      int minStartTime = currentStartTime;
+      while (it.hasNext()) {
+        Event nextEvent = it.next();
+        TimeRange nextEventTime = nextEvent.getWhen();
+        if (nextEventTime.end() >= currentStartTime) {
+          minStartTime = Math.min(currentStartTime, nextEventTime.start());
+        } else {
+          // In this case it is not possible to expand this event any farther,
+          // the iterator already points to the next element we should continue
+          // from.
+          // reset the iterator so the last tried value isn't skipped
+          it.previous();
+          break;
+        }
+      }
+
+      int fullDuration = currentEndTime - minStartTime;
+      // TODO(morleyd): Figure out why the TimeRange class constructor was
+      // defined as private, this seems confusing?
+      fullTimes.add(TimeRange.fromStartDuration(minStartTime, fullDuration));
+    }
+    System.out.print("The value of all the times before inversion is " +
+                     fullTimes);
+    // Finally we loop through the time in between these events (the inverse of
+    // the set), and if these times are within the length of the meeting we add
+    // them to our answer (going from smallest to biggest).
+    ListIterator<TimeRange> it2 = fullTimes.listIterator(fullTimes.size());
```

Comment:
Variables named `it` and `it2` are an immediate caution sign to me, suggesting that the naming or scoping could be improved. The names don't provide any semantic information, which may be fine when there is only a single, very-local iterator, but not when there are more or when they stick around for longer.

Also, I wonder if reversing the algorithm would allow you to omit this iterator entirely, since all it's doing is stepping through the list in reverse. (Or if Java has an easy way to get a reverse iterator; I was surprised not to see one at first glance.)

Reply:
I also found it surprising there was no easy reverse iterator option, you can add a specific apache library to get this functionality but I didn't think it was worth it for just this feature.

Reply:
I looked at these iterators and was a bit concerned, but then reasoned it wasn't much different than having an i and j index in a for loop. Is there a difference I am missing in this case?

Reply:
I agree it's not worth it to pull in a library for a reverse iterator.

I would argue that `i` and `j` as nested loop indices are often not the best choice either. If they're very local (e.g. the whole scope fits on one screen) and you don't really care about the semantics (e.g. you're indexing into a 2D array where the contents are arbitrary), maybe `i` and `j` are great. But even then, `row` and `col` might make mistakes easier to spot.

In this case, `it` and `it2` share a scope that's too big to fit on a screen (could be resolved by refactoring into smaller functions) and refer to semantically distinct things, the list of all events vs the list of busy times (could be resolved with more meaningful names). It would be easy to accidentally do something like `it.next()` instead of `it2.next()` in a context where you discard the return value, which could make for a fun bug.

Reply:
I'm working on the changes right now but I refactored the scope of it and it2 so they're declared in the scope of a for loop instead of outside a while. Would this make more sense?

Reply:
That's a fair point with it and it2 I've certainly had that bug before.

Reply:
Yes, it will likely be easier to read if each iterator is limited to inside a for loop (and they can both actually be `it`).